Wrappers/Python/ccpi/optimisation/operators/BlockScaledOperator.py

```diff
@@ -49,7 +49,7 @@ def adjoint(self, x, out=None):
             return self.scalar * self.operator.adjoint(x, out=out)
         else:
             raise TypeError('Operator is not linear')
-    def norm(self):
+    def calculate_norm(self):
```

Comment:
maybe this change is not needed as a `BlockScaledOperator` is inheriting from `ScaledOperator` which isn't a subclass of `Operator`.

Reply:
Not sure I understand this. In fact I don't think I am up to date on the relationships between the different block, scaled, linear (and non) operators.

Reply:
The `Operator` classes are laid out like this:

`object <- Operator <- LinearOperator`
`object <- Operator <- BlockOperator`
`object <- ScaledOperator`, though this implements all the methods in `Operator`
`object <- ScaledOperator <- BlockScaledOperator`

The `ScaledOperator` is not inheriting from `Operator` because it is the result of left multiplying an `Operator` with a scalar. We thought to implement it as a special case rather than having a scalar in the `Operator` and do a zillion of tests to see if the scalar is not 1 all the time any method is called. 

The docstring says

> ScaledOperator
    A class to represent the scalar multiplication of an Operator with a scalar.
    It holds an operator and a scalar. Basically it returns the multiplication
    of the result of direct and adjoint of the operator with the scalar.
    For the rest it behaves like the operator it holds.

Reply:
I see your point. Still I am wondering if this could be simplified slightly. Firstly, I may be overlooking something, but it appears to me that `ScaledOperator` could in fact inherit from `Operator` without any tests needed.  When ever the operator is called, Python will know if it is a `ScaledOperator` or `Operator`. In the former case, it will apply the scalar; in the latter, there is no scalar to apply.

Reply:
MMM, you might be right in facts!

Reply:
Second, I am wondering about how we handle linearity. We introduced `LinearOperator`, but then at least `ScaledOperator` has the attribute `is_linear`, so that it may be linear or not. Further, `BlockOperator` is inherently linear, as it has an adjoint and also uses a linear structure in how blocks are treated, however it is not a subclass of `LinearOperator`... I am wondering if we can get rid of `LinearOperator` and represent linearity across all operators with the attribute `is_linear` as done in `ScaledOperator`.

Reply:
Probably we'd need to implement the `direct` method for instance as

```python

def direct(self, x, out=None):
    return self.scalar * super(ScaledOperator, self).direct(x)
```

Let's keep this discussion for a next release.

Reply:
That of course might mean all the time checking whether `is_linear` is true as is already done in `ScaledOperator`.

Reply:
Well, `BlockOperator` has the method [`adjoint`](https://github.com/vais-ral/CCPi-Framework/blob/norm_refactor/Wrappers/Python/ccpi/optimisation/operators/BlockOperator.py#L164) which will fail if all the operators inside it are not `LinearOperator`s. 

Same with `ScaledOperator`, it [delegates](https://github.com/vais-ral/CCPi-Framework/blob/norm_refactor/Wrappers/Python/ccpi/optimisation/operators/ScaledOperator.py#L49) the linearity check to the operator it holds. 

To resume `is_linear` is present in all classes `Operator LinearOperator ScaledOperator BlockOperator` 

 * in `Operator` returns `False`. 
 * In `LinearOperator` it returns `True`. 
 * in `ScaledOperator` it delegates to the operator it contains
 * in `BlockOperator` it returns `True` only if all operators it holds are `LinearOperators`

I believe we introduced `LinearOperator` for compatibility with SIRF.

Reply:
I just found the [real reason](https://github.com/vais-ral/CCPi-Framework/issues/194#issuecomment-468404548) of the `ScaledOperator`. It is a SIRF/CIL compatibility.